File: app/src/main/java/ani/am/e_commerce/repositories/ProductRepository.java
package ani.am.e_commerce.repositories;

import android.arch.lifecycle.LiveData;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.JsonObject;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Executor;

import javax.inject.Singleton;

import ani.am.e_commerce.api.ApiInterface;
import ani.am.e_commerce.db.dao.CategoryDao;
import ani.am.e_commerce.db.dao.ProductDao;
import ani.am.e_commerce.db.entity.Category;
import ani.am.e_commerce.db.entity.Product;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static ani.am.e_commerce.activites.MainActivity.prefConfig;

@Singleton
public class ProductRepository {
    private final ApiInterface apiInterface;
    private final ProductDao productDao;
    private final Executor executor;

    public ProductRepository(ApiInterface apiInterface, ProductDao productDao, Executor executor) {
        this.apiInterface = apiInterface;
        this.productDao = productDao;
        this.executor = executor;
    }

    public void addProduct(String id, Map<String, RequestBody> map){
        String token = prefConfig.readToken("token");
        Log.d("Tag","add category " + id + "  token " + token);
        executor.execute(()->
                apiInterface.createProduct(id,token,map).enqueue(new Callback<JsonObject>() {
                    @Override
                    public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                        Log.d("Tag", response.toString());
                        if(response.body() != null) {
                            Log.d("Tag", response.body().toString());
                           // saveData(response.body().toString());
                        }

                    }

                    @Override
                    public void onFailure(Call<JsonObject> call, Throwable t) {

                    }
                }));
    }
    private void saveData(String json) {
        Log.d("Tag", "data json " + json);
        try {
            JSONObject jsonObject = new JSONObject(json);
            JSONObject data = jsonObject.getJSONObject("data");
            Gson gson = new Gson();
            Product product = gson.fromJson(data.toString(), Product.class);
            Log.d("Tag","updated Product " + product);
            productDao.insertProduct(product);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }
    /*public void updateProduct( Map<String, RequestBody> map){
        String token = prefConfig.readToken("token");
        String id = prefConfig.readToken("id");
        Log.d("Tag","update category " + id + "  token " + token);
        Log.d("Tag","update category " + map);
        executor.execute(()->
                apiInterface.updateCategory(id,token,map).enqueue(new Callback<JsonObject>() {
                    @Override
                    public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                        Log.d("Tag", response.toString());
                        if(response.body() != null)
                            Log.d("Tag", "updateCategory "+response.body().toString());
                    }

                    @Override
                    public void onFailure(Call<JsonObject> call, Throwable t) {
                        Log.d("Tag", "Fail updateCategory "+t.getMessage());
                    }
                }));
        //categoryDao.updateCategory();
    }*/

    public void deleteProduct(Product product) {
        String token = prefConfig.readToken("token");
        executor.execute(() -> {
                    apiInterface.deleteProduct(token, product.getId()).enqueue(new Callback<JsonObject>() {
                        @Override
                        public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                            Log.d("Tag", "Product deleted from network !" + response.message());
                        }

                        @Override
                        public void onFailure(Call<JsonObject> call, Throwable t) {

                        }
                    });
                   productDao.deleteProduct(product);
                }
        );
    }

    /*private void saveData(String json) {
        Log.d("Tag", "data json " + json);
        try {
            JSONObject jsonObject = new JSONObject(json);
            JSONArray data = jsonObject.getJSONArray("data");
            List<Category> categoryList = new ArrayList<>();
            Gson gson = new Gson();
            for (int i = 0; i < data.length(); i++) {
                JSONObject object = data.getJSONObject(i);
                Category category = gson.fromJson(object.toString(), Category.class);
                categoryList.add(category);
            }
            Log.d("Tag", "categorylist size before save " + categoryList.size());
            categoryDao.saveList(categoryList);

        } catch (JSONException e) {
            e.printStackTrace();
        }
    }*/

}

File: app/src/main/java/ani/am/e_commerce/repositories/ProductRepository.kt
package ani.am.e_commerce.repositories

import android.arch.lifecycle.LiveData
import android.location.Criteria
import android.util.Log
import ani.am.e_commerce.activities.MainActivity
import ani.am.e_commerce.api.ApiInterface
import ani.am.e_commerce.db.dao.ProductDao
import ani.am.e_commerce.db.entity.Product
import com.google.gson.Gson
import com.google.gson.JsonObject
import okhttp3.RequestBody
import org.json.JSONException
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.concurrent.Executor
import javax.inject.Singleton

@Singleton
class ProductRepository(
        private val apiInterface: ApiInterface,
        private val productDao: ProductDao,
        private val executor: Executor) {

    fun addProduct(id: String, map: Map<String, RequestBody>) {
        val token = MainActivity.prefConfig.readToken("token")
        Log.d("Tag", "add product $id  token $token")
        executor.execute {
            apiInterface.createProduct(id, token, map).enqueue(object : Callback<JsonObject> {
                override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
                    executor.execute {
                        Log.d("Tag", response.toString())
                        if (response.body() != null) {
                            Log.d("Tag", response.body().toString())
                            saveData(response.body().toString())
                        }
                    }
                }

                override fun onFailure(call: Call<JsonObject>, t: Throwable) {}
            })
        }
    }

    fun updateProduct(id: String, map: Map<String, RequestBody>) {
        val token = MainActivity.prefConfig.readToken("token")
        Log.d("Tag", "update product $id  token $token")
        Log.d("Tag", "update product $map")
        executor.execute {
            apiInterface.updateProduct( id, token, map).enqueue(object : Callback<JsonObject> {
                override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
                    Log.d("Tag", response.toString())
                    if (response.body() != null) {
                        Log.d("Tag", "updateProduct " + response.body().toString())
                        updateData(response.body().toString());
                    }
                }

                override fun onFailure(call: Call<JsonObject>, t: Throwable) {
                    Log.d("Tag", "Fail updateProduct " + t.message)
                }
            })
        }
    }

    fun saveData(json: String) {
        executor.execute {
            Log.d("Tag", "data json $json")
            try {
                val jsonObject = JSONObject(json)
                val data = jsonObject.getJSONObject("data")
                val gson = Gson()
                val product = gson.fromJson(data.toString(), Product::class.java)
                Log.d("Tag", "updated Product $product")
                productDao.insertProduct(product)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

    fun updateData(json: String) {
        executor.execute {
            Log.d("Tag", "data json $json")
            try {
                val jsonObject = JSONObject(json)
                val data = jsonObject.getJSONObject("data")
                val gson = Gson()
                val product = gson.fromJson(data.toString(), Product::class.java)
                Log.d("Tag", "updated Product $product")
                productDao.updateProduct(product)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

    fun deleteProduct(product: Product) {
        val token = MainActivity.prefConfig.readToken("token")
        executor.execute {
            apiInterface.deleteProduct(token, product.id).enqueue(object : Callback<JsonObject> {
                override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
                    Log.d("Tag", "Product deleted from network !" + response.message())
                    if (response.isSuccessful)
                        executor.execute {
                            productDao.deleteProduct(product)
                        }
                }

                override fun onFailure(call: Call<JsonObject>, t: Throwable) {}
            })
            productDao.deleteProduct(product)
        }
    }

    fun searchProduct(criteria: String){
        executor.execute(){
            apiInterface.searchProduct(criteria).enqueue(object :Callback<JsonObject>{
                override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
                    Log.d("Tag", "Product searched!" + response.message())
                    if (response.isSuccessful)
                        executor.execute {
                            saveProductsList(response.body().toString())
                        }
                }

                override fun onFailure(call: Call<JsonObject>, t: Throwable) {}

            })
        }
    }

    fun getSearchableList (criteria: String):LiveData<List<Product>>{
        searchProduct(criteria)
        return productDao.allProducts
    }

    fun getAllProductsByCategoryId(id: String): LiveData<List<Product>> {
        refreshProductsListByCategoryId(id)
        return productDao.getAllProductsByCategoryId(id)
    }

    fun refreshProductsListByCategoryId(categoryId: String) {
        //val token = MainActivity.prefConfig.readToken("token")
        executor.execute {
            apiInterface.getAllProductByCategoryId(categoryId).enqueue(object : Callback<JsonObject> {
                override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
                    Log.d("Tag", "refreshProductsListByCategoryId $response.toString()")
                    if (response.body() != null) {
                        Log.d("Tag", response.body().toString())
                        saveProductsList(response.body().toString())
                    }
                }

                override fun onFailure(call: Call<JsonObject>, t: Throwable) {}
            })
        }
    }

    fun saveProductsList(json: String) {
        val listProduct = arrayListOf<Product>()
        executor.execute {
            Log.d("Tag", "data json $json")
            val jsonObject = JSONObject(json)
            val data = jsonObject.getJSONArray("data")
            try {
                for (i in 0..data.length() - 1) {
                    val productObject = data.getJSONObject(i)
                    val gson = Gson()
                    val product = gson.fromJson(productObject.toString(), Product::class.java)
                    listProduct.add(product)
                }

                Log.d("Tag", "updated Product $listProduct")
                productDao.saveProductList(listProduct)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
    }

}